### PR TITLE
[DowngradePHP74] Fix parent removal regression

### PR DESCRIPTION
**Pull request by TomasVotruba**
Fixed regression in `parent` type resolution in https://github.com/rectorphp/rector-src/pull/1071

Closes https://github.com/rectorphp/rector/issues/6772 bug

### CODE REVIEW

**Comment by samsonasik** [comment]
@TomasVotruba this is same with failing PR https://github.com/rectorphp/rector-src/pull/1080 . The use case is different, as in the `rector` build, the `: parent` actually removed entirely, but it is probably related and need to be skipped.

**Comment by TomasVotruba** [comment]
> the : parent actually removed entirely

I think this happens in cascade by the next rule taking over alraedy broken code. 
I'll see how can I fix this one first.

**Comment by samsonasik** [comment]
The `\Symfony\Component\String\AbstractString::class,` may need to be removed from `build/config/config-downgrade.php` like in PR https://github.com/rectorphp/rector-src/pull/1080/files#diff-c8125c36a958b829ffd79f1774dc9649d2ef21b044e4d849225960bfed613a80

File: packages/StaticTypeMapper/PhpParser/NameNodeMapper.php [hunk 1]
```diff
@@ -83,7 +85,12 @@ private function createClassReferenceType(Name $name, string $reference): MixedT
         }
 
         if ($reference === ObjectReference::PARENT()->getValue()) {
-            return new ParentStaticType($classReflection);
+            $parentClassReflection = $classReflection->getParentClass();
+            if (! $parentClassReflection instanceof ClassReflection) {
+                throw new ShouldNotHappenException();
+            }
+
+            return new ParentStaticType($parentClassReflection);
```
Comment (TomasVotruba) [inline packages/StaticTypeMapper/PhpParser/NameNodeMapper.php]: @samsonasik Here was the bug probably. The `parent` did not refere to parent class, but self class.

```diff
-return new ParentStaticType($classReflection);
+return new ParentStaticType($parentClassReflection);
```

Comment (samsonasik) [reply packages/StaticTypeMapper/PhpParser/NameNodeMapper.php]: @TomasVotruba thank you 👍

**Comment by samsonasik** [comment]
@TomasVotruba I tried locally, it seems still not working, with different error:

```bash
➜  rector-src git:(tv-keep-parent-reference) ✗ cd rector-prefixed-downgraded 
➜  rector-prefixed-downgraded git:(tv-keep-parent-reference) ✗ cp ../build/target-repository/bootstrap.php .
➜  rector-prefixed-downgraded git:(tv-keep-parent-reference) ✗ cp ../preload.php .
➜  rector-prefixed-downgraded git:(tv-keep-parent-reference) ✗ bin/rector list --ansi
Rector 2215b0e85570ca53d450bf83c20cf3629e3d7680

Usage:
  command [options] [arguments]

PHP Fatal error:  Declaration of RectorPrefix20211027\Symfony\Component\String\UnicodeString::join($strings, $lastGlue = null) must be compatible with RectorPrefix20211027\Symfony\Component\String\AbstractUnicodeString::join($strings, $lastGlue = null): RectorPrefix20211027\Symfony\Component\String\AbstractString in /Users/samsonasik/www/rector-src/rector-prefixed-downgraded/vendor/symfony/string/UnicodeString.php on line 169
Fatal error: Declaration of RectorPrefix20211027\Symfony\Component\String\UnicodeString::join($strings, $lastGlue = null) must be compatible with RectorPrefix20211027\Symfony\Component\String\AbstractUnicodeString::join($strings, $lastGlue = null): RectorPrefix20211027\Symfony\Component\String\AbstractString in /Users/samsonasik/www/rector-src/rector-prefixed-downgraded/vendor/symfony/string/UnicodeString.php on line 169
```

**Comment by samsonasik** [comment]
it still error in CI build as well https://github.com/rectorphp/rector-src/runs/4021163113?check_suite_focus=true#step:18:11

**Comment by samsonasik** [comment]
@TomasVotruba the `AbstractString::join()` return self seems need to be kept, and the child `UnicodeString::join()` return AbstractString seems need to be kept as well.

**Pull request merged**